onboard/nucleo_bridge/Controller.cpp

```diff
@@ -0,0 +1,166 @@
+#include "Controller.h"
+
+//Wrapper for backend->i2c_transact, autofilling the i2c address of the Controller
+void Controller::transact(uint8_t cmd, uint8_t writeNum, uint8_t readNum, uint8_t *writeBuf, uint8_t *readBuf)
+{
+    I2C::transact(ControllerMap::get_i2c_address(name), cmd, writeNum, readNum, writeBuf, readBuf);
+}
+
+//If this Controller is not live, make it live by configuring the real controller
+void Controller::make_live()
+{
+    if (ControllerMap::check_if_live(name))
+    {
+        return;
+    }
+
+    try
+    {
+        uint8_t buffer[32];
+        memcpy(buffer, POINTER(&(hardware.pwmMin)), 2);
+        memcpy(buffer + 2, POINTER(&(hardware.pwmMax)), 2);
+        memcpy(buffer + 4, POINTER(&(hardware.pwmPeriod)), 2);
+        transact(CONFIG_PWM, buffer, nullptr);
+
+        memcpy(buffer, POINTER(&(kP)), 4);
+        memcpy(buffer + 4, POINTER(&(kI)), 4);
+        memcpy(buffer + 8, POINTER(&(kD)), 4);
+        transact(CONFIG_K, buffer, nullptr);
+
+        uint16_t input = 0;
+        //transact(SPI, nullptr, POINTER(&input));
+
+        int32_t angle = static_cast<int32_t>(quadCPR * ((static_cast<float>(input) / spiCPR) + (startAngle / (2.0 * M_PI))));
+        transact(ADJUST, POINTER(&angle), nullptr);
+
+        transact(ON, nullptr, nullptr);
+
+        ControllerMap::make_live(name);
+    }
+    catch (IOFailure &e)
+    {
+        printf("activate failed on %s\n", name.c_str());
+        throw IOFailure();
+    }
+}
+
+//Helper function to convert raw angle to radians. Also checks if new angle is close to old angle
+void Controller::recordAngle(int32_t angle)
+{
+    float otherAngle = (static_cast<float>(angle) / quadCPR) * 2.0 * M_PI;
+    if (std::abs(otherAngle - currentAngle) >= M_PI / 16.0)
+    {
+        return;
+    }
+    currentAngle = otherAngle;
+}

```

Comment:
Just curious, are early returns important

Reply:
It's considered better style to avoid early returns. They make the code less maintainable and more prone to bugs. As the function grows, people will have to find more return statements if they need to do something at the end of the function. Early returns could be missed easily.